File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/storage/SelectPresenter.java
package app.editors.manager.mvp.presenters.storage;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import app.editors.manager.R;
import app.editors.manager.app.Api;
import app.editors.manager.app.App;
import app.editors.manager.managers.tools.AccountSqlTool;
import app.editors.manager.managers.tools.PreferenceTool;
import app.editors.manager.managers.tools.RetrofitTool;
import app.editors.manager.managers.utils.StorageUtils;
import app.editors.manager.mvp.models.account.Storage;
import app.editors.manager.mvp.views.storage.SelectView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public class SelectPresenter extends MvpPresenter<SelectView> {

    public static final String TAG = SelectPresenter.class.getSimpleName();

    @Inject
    Context mContext;
    @Inject
    PreferenceTool mPreferenceTool;
    @Inject
    RetrofitTool mRetrofitTool;
    @Inject
    AccountSqlTool mAccountSqlTool;

    private CompositeDisposable mDisposable;
    private List<Storage> mStorages;

    public SelectPresenter() {
        App.getApp().getAppComponent().inject(this);
        mDisposable = new CompositeDisposable();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mDisposable != null) {
            mDisposable.clear();
        }
    }

    public void getStorages() {
        getViewState().showProgress(true);
        mDisposable.add(mRetrofitTool.getApiWithPreferences().getThirdpartyCapabilities(mPreferenceTool.getToken())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseBody -> new JSONObject(responseBody.string()))
                .map(this::collectListStorage)
                .subscribe(list -> {
                    mStorages = list;
                    getViewState().onUpdate(getNames(list));
                    getViewState().showProgress(false);
                }, error -> {
                    getViewState().onError(error.getMessage());
                }));
    }

    private ArrayList<Storage> collectListStorage(JSONObject jsonObject) throws JSONException {
        ArrayList<Storage> storages = new ArrayList<>();
        JSONArray array = jsonObject.getJSONArray("response");

        for (int i = 0; i < array.length(); i++) {
            JSONArray arrayString = array.getJSONArray(i);
            Storage storage = new Storage();

            for (int j = 0; j < arrayString.length(); j++) {
                if (j == 0) {
                    storage.setName(arrayString.getString(j));
                } else if (j == 1) {
                    storage.setClientId(arrayString.getString(j));
                } else if (j == 2) {
                    storage.setRedirectUrl(arrayString.getString(j));
                }

            }
            storages.add(storage);
        }
        return storages;
    }

    public void connect(String providerKey) {
        Storage storage = getStorage(providerKey);

        if (storage != null && storage.getClientId() != null) {
            getViewState().showWebTokenFragment(storage);
        } else {
            String url = "";
            String title = "WebDav service";
            switch (providerKey) {
                case Api.Storage.YANDEX:
                    url = StorageUtils.WevDav.URL_YANDEX;
                    title = mContext.getString(R.string.storage_select_yandex);
                    break;
                case Api.Storage.SHAREPOINT:
                    title = mContext.getString(R.string.storage_select_share_point);
                    break;
                case Api.Storage.OWNCLOUD:
                    title = mContext.getString(R.string.storage_select_own_cloud);
                    providerKey = Api.Storage.WEBDAV;
                    break;
                case Api.Storage.NEXTCLOUD:
                    title = mContext.getString(R.string.storage_select_next_cloud);
                    providerKey = Api.Storage.WEBDAV;
                    break;
            }

            getViewState().showWebDavFragment(providerKey, url, title);
        }

    }

    private Storage getStorage(String providerKey) {
        if (mStorages != null) {
            for (Storage storage : mStorages) {
                if (storage.getName().equals(providerKey)) {
                    return storage;
                }
            }
        }
        return null;
    }

    private List<String> getNames(ArrayList<Storage> list) {
        List<String> names = new ArrayList<>();
        for (Storage storage : list) {
            names.add(storage.getName());
        }
        return names;
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/storage/SelectPresenter.kt
package app.editors.manager.mvp.presenters.storage

import android.content.Context
import app.documents.core.storage.account.AccountDao
import app.documents.core.network.common.contracts.ApiContract
import app.editors.manager.R
import app.editors.manager.app.Api
import app.editors.manager.app.App
import app.editors.manager.app.api
import app.editors.manager.managers.utils.StorageUtils
import app.editors.manager.mvp.models.account.Storage
import app.editors.manager.mvp.views.storage.SelectView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import moxy.InjectViewState
import moxy.MvpPresenter
import okhttp3.ResponseBody
import org.json.JSONException
import org.json.JSONObject
import javax.inject.Inject

@InjectViewState
class SelectPresenter : MvpPresenter<SelectView>() {

    companion object {
        val TAG: String = SelectPresenter::class.java.simpleName
    }

    @Inject
    lateinit var context: Context

    @Inject
    lateinit var accountDao: AccountDao

    init {
        App.getApp().appComponent.inject(this)
    }

    private var disposable: Disposable? = null
    private var storageList: List<Storage>? = null

    private val api: Api = context.api()

    override fun onDestroy() {
        super.onDestroy()
        disposable?.dispose()
    }

    fun getStorages() {
        viewState.showProgress(true)
        disposable = api.thirdpartyCapabilities()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseBody: ResponseBody -> JSONObject(responseBody.string()) }
            .map { jsonObject: JSONObject -> collectListStorage(jsonObject) }
            .subscribe({ list: ArrayList<Storage> ->
                storageList = list
                viewState.onUpdate(getNames(list))
                viewState.showProgress(false)
            }, { error: Throwable -> viewState.onError(error.message) })
    }

    @Throws(JSONException::class)
    private fun collectListStorage(jsonObject: JSONObject): ArrayList<Storage> {
        val storages = ArrayList<Storage>()
        val nextcloud = Storage(name = ApiContract.Storage.NEXTCLOUD)
        val owncloud = Storage(name = ApiContract.Storage.OWNCLOUD)
        val kDrive = Storage(name = ApiContract.Storage.KDRIVE)
        val sharePoint = Storage(name = ApiContract.Storage.SHAREPOINT)
        val yandex = Storage(name = ApiContract.Storage.YANDEX)
        val webDav = Storage(name = ApiContract.Storage.WEBDAV)
        storages.add(nextcloud)
        storages.add(owncloud)
        storages.add(kDrive)
        storages.add(sharePoint)
        val array = jsonObject.getJSONArray("response")
        for (i in 0 until array.length()) {
            val arrayString = array.getJSONArray(i)
            if (arrayString.length() == 3) {
                val storage = Storage()
                for (j in 0 until arrayString.length()) {
                    when (j) {
                        0 -> {
                            storage.name = arrayString.getString(j)
                        }
                        1 -> {
                            storage.clientId = arrayString.getString(j)
                        }
                        2 -> {
                            storage.redirectUrl = arrayString.getString(j)
                        }
                    }
                }
                storages.add(storage)
            }
        }
        storages.add(yandex)
        storages.add(webDav)
        return storages
    }

    fun connect(providerKey: String) {
        var key = providerKey
        val storage = getStorage(key)
        if (storage?.clientId != null) {
            viewState.showWebTokenFragment(storage)
        } else {
            var url = ""
            var title = "WebDav service"
            when (key) {
                ApiContract.Storage.YANDEX -> {
                    url = StorageUtils.WevDav.URL_YANDEX
                    title = context.getString(R.string.storage_select_yandex)
                }
                ApiContract.Storage.KDRIVE -> {
                    url = StorageUtils.WevDav.URL_KDRIVE
                    title = context.getString(R.string.storage_select_kdrive)
                }
                ApiContract.Storage.SHAREPOINT -> {
                    title = context.getString(R.string.storage_select_share_point)
                    key = ApiContract.Storage.WEBDAV
                }
                ApiContract.Storage.OWNCLOUD -> {
                    title = context.getString(R.string.storage_select_own_cloud)
                    key = ApiContract.Storage.WEBDAV
                }
                ApiContract.Storage.NEXTCLOUD -> {
                    title = context.getString(R.string.storage_select_next_cloud)
                    key = ApiContract.Storage.WEBDAV
                }
            }
            viewState.showWebDavFragment(key, url, title)
        }
    }

    private fun getStorage(providerKey: String): Storage? {
        if (storageList != null) {
            for (storage in storageList!!) {
                if (storage.name == providerKey) {
                    return storage
                }
            }
        }
        return null
    }

    private fun getNames(list: ArrayList<Storage>): List<String> {
        val names: MutableList<String> = ArrayList()
        for (storage in list) {
            names.add(storage.name ?: "")
        }
        return names
    }
}